File: strategy/marvin/task/defendpenalty.lua
local DefendPenalty = Class("Task.DefendPenalty", require "task/base")

local geom = require "../base/geom"
local vis = require "../base/vis"
local World = require "../base/world"
local PathHelper = require "trajectory/pathhelper"
local ToTarget = require "trajectory/totarget"
local Interval = require "util/interval"


local PENALTY_LINE_DISTANCE = 0.35 -- prevent robots from crossing the penalty line

local obstacleTable = {
	ignoreBall = false,
	ignorePass = true
}


function DefendPenalty:run()
	local rr = self._robot.radius --assume all robots have the same radius
	local penaltyLine = World.Geometry.OwnPenaltyLine + PENALTY_LINE_DISTANCE
	vis.addPath("t/defendpenalty: penaltyDistance", {Vector(-2,penaltyLine), Vector(2,penaltyLine)}, vis.colors.whiteHalf)
	-- NOTE: All spots are on the penaltyline, so only x-values are processed

	local occupiedSpotsFriendly = {}
	for robot, pos in pairs(self._inbox.moveDest()) do
		if math.abs(pos.y - penaltyLine) < 2*rr and robot.id > self._robot.id then
			table.insert(occupiedSpotsFriendly, pos.x)
		end
	end

	local occupiedSpotsOpp = {} -- positions of opponents on the line
	for _, robot in ipairs(World.OpponentRobots) do
		if math.abs(robot.pos.y - penaltyLine) < rr then
			table.insert(occupiedSpotsOpp, robot.pos.x)
		end
	end
	local preferredSpots = {}
	for _, robot in ipairs(World.OpponentRobots) do
		if robot ~= World.OpponentKeeper and (robot.pos.y+rr) > penaltyLine then
			-- prefer spot between own keeper and opponent to catch rebound
			local ownKeeperPos = World.FriendlyKeeper and World.FriendlyKeeper.pos or World.Geometry.FriendlyGoal
			local keeperOppDir = robot.pos - ownKeeperPos
			local prefSpot = (geom.intersectLineLine(ownKeeperPos, keeperOppDir, Vector(0, penaltyLine), Vector.fromAngle(math.pi)))
			if prefSpot then
				table.insert(preferredSpots, prefSpot.x)
			end
		end
	end

	local targetPos
	-- preference one: next to an opponent on the penaltyLine
	table.sort(occupiedSpotsOpp)
	for i = 1, #occupiedSpotsOpp do
		-- ignore if other defender is there
		local alreadyMarked = false
		for _, defX in ipairs(occupiedSpotsFriendly) do
			if math.abs(occupiedSpotsOpp[i] - defX) < 2.5* rr then
				alreadyMarked = true
			end
		end
		if not alreadyMarked then
			-- check dist to next occupied spot
			local left = occupiedSpotsOpp[i-1] and math.abs(occupiedSpotsOpp[i-1] - occupiedSpotsOpp[i]) < 2.5* rr
			local right = occupiedSpotsOpp[i+1] and math.abs(occupiedSpotsOpp[i+1] - occupiedSpotsOpp[i]) < 2.5* rr
			local leftPos = occupiedSpotsOpp[i] - 2*rr
			local rightPos = occupiedSpotsOpp[i] + 2*rr
			-- prefer side to the middle
			if occupiedSpotsOpp[i] > 0 then -- opponent is on the right side
				if not left then
					targetPos = leftPos
					break
				elseif not right then
					targetPos = rightPos
					break
				end
			else -- opponent is on the left side
				if not right then
					targetPos = rightPos
					break
				elseif not left then
					targetPos = leftPos
					break
				end
			end
		end
	end

	local occupiedSpotsAll = table.combine(occupiedSpotsOpp, occupiedSpotsFriendly)
	if not targetPos then -- preference two: intersection of penaltyLine and line from opponent to friendlyKeeper
		for _, prefX in ipairs(preferredSpots) do
			local noOneNear = true
			for _, occX in ipairs(occupiedSpotsAll) do
				if math.abs(prefX - occX) < 2*rr then
					noOneNear = false
					break
				end
			end
			if noOneNear then
				targetPos = prefX
			end
		end
	end
	if not targetPos then -- fallback: search free point on penaltyLine, which is closest to the middle
		local occupiedSectors = table.map(occupiedSpotsAll, function(x) return {x-rr,x+rr} end)
		Interval.sort(occupiedSectors)
		Interval.merge(occupiedSectors)
		local widthLimit = World.Geometry.FieldWidthHalf - 2 * self._robot.radius
		local freeSectors = Interval.negate(occupiedSectors, -widthLimit, widthLimit)
		targetPos = Interval.getClosestPoint(freeSectors, 0, rr)
	end

	if not targetPos then --should only occur when all the whole penalty line is full with robots (i.e never)
		targetPos = Vector(0, 0)
	else
		targetPos = Vector(targetPos, penaltyLine)
	end

	PathHelper.setDefaultObstacles(self._robot.path, self._robot, obstacleTable)
	self._robot.trajectory:update(ToTarget, targetPos, (World.Ball.pos - self._robot.pos):angle())

	self._send.moveDest("all", targetPos)
end

return DefendPenalty

File: strategy/typescript/glados/task/defender/defendpenalty.ts
import * as geom from "base/geom";
import * as vis from "base/vis";
import {Vector, Position} from "base/vector";
import * as World from "base/world";
import {MessageType} from "glados/control/messaging";
import * as PathHelper from "glados/trajectory/pathhelper";
import {ToTarget} from "glados/trajectory/totarget";
import * as Interval from "glados/util/interval";
import {Task} from "glados/task/base";


let PENALTY_LINE_DISTANCE = 0.35 // prevent robots from crossing the penalty line

let obstacleTable: PathHelper.PathHelperParameters = {
	ignoreBall: false,
	ignorePass: true
}

export class DefendPenalty extends Task {
	run () {
		let rr = this._robot.radius; //assume all robots have the same radius
		let penaltyLine = World.Geometry.OwnPenaltyLine + PENALTY_LINE_DISTANCE;
		vis.addPath("t/defendpenalty: penaltyDistance", [new Vector(-2,penaltyLine), new Vector(2,penaltyLine)], vis.colors.whiteHalf);
		// NOTE: All spots are on the penaltyline, so only x-values are processed

		let occupiedSpotsFriendly: number[] = [];
		for (let [robot, pos] of this._messaging.receive(MessageType.moveDest)) {
			if (Math.abs(pos.y - penaltyLine) < 2*rr && robot.id > this._robot.id) {
				occupiedSpotsFriendly.push(pos.x);
			}
		}

		let occupiedSpotsOpp: number[] = []; // positions of opponents on the line
		for (let robot of World.OpponentRobots) {
			if (Math.abs(robot.pos.y - penaltyLine) < rr) {
				occupiedSpotsOpp.push(robot.pos.x);
			}
		}
		let preferredSpots: number[] = [];
		for (let robot of World.OpponentRobots) {
			if (robot != World.OpponentKeeper && (robot.pos.y+rr) > penaltyLine) {
				// prefer spot between own keeper and opponent to catch rebound
				let ownKeeperPos = World.FriendlyKeeper ? World.FriendlyKeeper.pos : World.Geometry.FriendlyGoal
				let keeperOppDir = robot.pos - ownKeeperPos
				let prefSpot = geom.intersectLineLine(ownKeeperPos, keeperOppDir, new Vector(0, penaltyLine), Vector.fromAngle(Math.PI))[0];
				if (prefSpot) {
					preferredSpots.push(prefSpot.x);
				}
			}
		}

		let targetPos: number | undefined;
		// preference one: next to an opponent on the penaltyLine
		occupiedSpotsOpp.sort();
		for (let i = 0;i<occupiedSpotsOpp.length;i++) {
			// ignore if other defender is there
			let alreadyMarked = false;
			for (let defX of occupiedSpotsFriendly) {
				if (Math.abs(occupiedSpotsOpp[i] - defX) < 2.5* rr) {
					alreadyMarked = true;
				}
			}
			if (!alreadyMarked) {
				// check dist to next occupied spot
				let left = i > 0 && Math.abs(occupiedSpotsOpp[i-1] - occupiedSpotsOpp[i]) < 2.5* rr;
				let right = i < occupiedSpotsOpp.length-1 && Math.abs(occupiedSpotsOpp[i+1] - occupiedSpotsOpp[i]) < 2.5* rr;
				let leftPos = occupiedSpotsOpp[i] - 2*rr;
				let rightPos = occupiedSpotsOpp[i] + 2*rr;
				// prefer side to the middle
				if (occupiedSpotsOpp[i] > 0) { // opponent is on the right side
					if (!left) {
						targetPos = leftPos;
						break;
					} else if (!right) {
						targetPos = rightPos;
						break;
					}
				} else {// opponent is on the left side
					if (!right) {
						targetPos = rightPos;
						break;
					} else if (!left) {
						targetPos = leftPos;
						break;
					}
				}
			}
		}

		let occupiedSpotsAll = occupiedSpotsOpp.concat(occupiedSpotsFriendly);
		if (targetPos == undefined) { // preference two: intersection of penaltyLine and line from opponent to friendlyKeeper
			for (let prefX of preferredSpots) {
				let noOneNear = true;
				for (let occX of occupiedSpotsAll) {
					if (Math.abs(prefX - occX) < 2*rr) {
						noOneNear = false;
						break;
					}
				}
				if (noOneNear) {
					targetPos = prefX;
				}
			}
		}
		if (targetPos == undefined) { // fallback: search free point on penaltyLine, which is closest to the middle
			let occupiedSectors = occupiedSpotsAll.map(function(x: number): [number, number] { return [x-rr,x+rr]; })
			Interval.sort(occupiedSectors);
			Interval.merge(occupiedSectors);
			let widthLimit = World.Geometry.FieldWidthHalf - 2 * this._robot.radius;
			let freeSectors = Interval.negate(occupiedSectors, -widthLimit, widthLimit);
			targetPos = Interval.getClosestPoint(freeSectors, 0, rr);
		}

		let resultPos: Position;
		if (targetPos == undefined) { //should only occur when all the whole penalty line is full with robots (i.e never)
			resultPos = new Vector(0, 0);
		} else {
			resultPos = new Vector(targetPos, penaltyLine);
		}

		PathHelper.setDefaultObstaclesByTable(this._robot.path, this._robot, obstacleTable)
		this._robot.trajectory.update(ToTarget, resultPos, (World.Ball.pos - this._robot.pos).angle())

		this._messaging.sendBroadcast(MessageType.moveDest, resultPos);
	}
}